File: mavkit-python/examples/monitor_link_state.py
import asyncio
import os

import mavkit


def is_terminal(state: mavkit.LinkState) -> bool:
    return state in (mavkit.LinkState.Disconnected, mavkit.LinkState.Error)


async def main():
    bind_addr = os.environ.get("MAVKIT_EXAMPLE_UDP_BIND", "0.0.0.0:14550")

    async with await mavkit.Vehicle.connect_udp(bind_addr) as vehicle:
        state_handle = vehicle.link().state()
        current = state_handle.latest()

        if current is not None:
            print(f"link state: {current}")
            if is_terminal(current):
                return

        subscription = state_handle.subscribe()
        last_state = current

        while True:
            state = await subscription.recv()
            if state == last_state:
                continue

            print(f"link state: {state}")
            if is_terminal(state):
                return

            last_state = state


asyncio.run(main())

File: examples/monitor_link_state.rs
use mavkit::{LinkState, Vehicle};

fn is_terminal(state: &LinkState) -> bool {
    matches!(state, LinkState::Disconnected | LinkState::Error(_))
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let bind_addr =
        std::env::var("MAVKIT_EXAMPLE_UDP_BIND").unwrap_or_else(|_| "0.0.0.0:14550".to_string());

    let vehicle = Vehicle::connect_udp(&bind_addr).await?;
    let state_handle = vehicle.link().state();

    if let Some(state) = state_handle.latest() {
        println!("link state: {state:?}");
        if is_terminal(&state) {
            return Ok(());
        }
    }

    let mut subscription = state_handle.subscribe();
    let mut last_state = state_handle.latest();

    while let Some(state) = subscription.recv().await {
        if last_state.as_ref() == Some(&state) {
            continue;
        }

        println!("link state: {state:?}");
        if is_terminal(&state) {
            break;
        }

        last_state = Some(state);
    }

    Ok(())
}
